Adding glassdoor get_jobs function

ofeksr opened the pull request:
My first pull request ever (as you can see?)

ofeksr committed:
glassdoor.get_jobs

Avielyo10 commented on .idea/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
+/workspace.xml
```
Move `.gitignore` outside of `.idea/` directory

ofeksr replied:
Created automatically .. deleted

Avielyo10 replied:
You should delete it again, perhaps put `.idea/` directory inside `.gitignore`

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -1,6 +1,11 @@
+import random
+import time
+
 import requests
```
Put all the `import` together then space and then `from x import y`  together

ofeksr replied:
PyCharm optimize imports function did that

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+        :param job_type: optional, 'fulltime' or ___ or ___ TODO: get job_type possibilities
+        :param job_location: Jerusalem (Israel), Tel Aviv-Yafo (Israel), Haifa (Israel), Ashdod (Israel),
+         Bnei Brak (Israel), Netanya (Israel), Bene Beraq (Israel), Bat Yam (Israel), Ramat Gan(Israel),
+         Ashqelon (Israel).
```
Can't we have `any`?

ofeksr replied:
I didn't find an option to search in 'Israel' , so by default if you don't set 'job_location' function parameter you will get all the jobs lists one after another for these locations.

ofeksr replied:
Got list of all cities, by default job location is None so it will show general israel jobs.

Avielyo10 replied:
We want to present jobs in all of the cities in Israel by default

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+                   'user-agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)'
+                                 ' Ubuntu Chromium/51.0.2704.79 Chrome/51.0.2704.79 Safari/537.36',
+                   'Cache-Control': 'no-cache',
+                   'Connection': 'keep-alive'
```
I think those lines are not needed, please try to test without those lines

ofeksr replied:
You are right, only user-agent needed in both headers.

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+            'user-agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)'
+                          ' Ubuntu Chromium/51.0.2704.79 Chrome/51.0.2704.79 Safari/537.36',
+            'Cache-Control': 'no-cache',
+            'Connection': 'keep-alive'
```
See above

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+            location_response = [dictionary for dictionary in location_response
+                                 if job_location in dictionary['label']]
+        else:
+            print('Job location not found, printing jobs from all locations in Israel..')
```
We want to get all the jobs in Israel by default, filtering by a location is an extra step

ofeksr replied:
Found a way to do that, searched a little deeper in the post request, but location need to be exact match as in "glassdoor israel cities" file that i'll add.

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+        else:
+            print('Job location not found, printing jobs from all locations in Israel..')
+
+        for i, location in enumerate(location_response):
```
Use `index` instead of `i`

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+                if i > 0:
+                    delays = [3, 2, 1]
+                    delay = random.choice(delays)
+                    time.sleep(delay)
```
Great idea!
Replace lines `101-102` with `delay = random.randrange(1,4)`

Avielyo10 replied:
Or simply:
```python
if index > 0:
    time.sleep(random.randrange(1,4))
```

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+                                        job_listings[place_label][job_id].append(a.string)
+                            else:
+                                job_listings[place_label] = {job_id: [a.string, full_link]}
+
```
Maybe use `switch case statement` instead of all of those nested `if`s
[How to implement a switch-case statement in Python?](https://jaxenter.com/implement-switch-case-statement-python-138315.html)

Avielyo10 replied:
```python
    if a['href'].startswith('/partner/') and \
        (a.string is not None) and \
        (a.string != 'JobInfo'):

        link = a['href']
        full_link = 'https://www.glassdoor.com'+link 
        job_id = link.split('jobListingId=')[1]
        if (place_label in job_listings) and (job_id not in job_listings[place_label]):
             job_listings[place_label].update({job_id: [a.string, full_link]})
        elif (place_label in job_listings) and (a.string not in job_listings[place_label][job_id]):
             job_listings[place_label][job_id].append(a.string)
        else:
            job_listings[place_label] = {job_id: [a.string, full_link]}
```
Or using this snippet

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+                                if job_id not in job_listings[place_label]:
+                                    job_listings[place_label].update({job_id: [a.string, full_link]})
+                                else:
+                                    if a.string not in job_listings[place_label][job_id]:
```
`elif`?

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -34,6 +39,99 @@ class Glassdoor(Resource):
+
+                    if a['href'].startswith('/partner/'):
+                        link = a['href']
+                        full_link = 'https://www.glassdoor.com'+link  # TODO: shrink full links length with tiny url?
```
We don't have to shrink it, full link is totally OK :)

ofeksr committed:
example for usage:
```
g = Glassdoor()
    g.get_jobs(job_title='Student', job_location='Tel Aviv', job_type='parttime')
```

ofeksr commented on legendary_eureka.py:
```diff
@@ -36,102 +33,120 @@ def get_jobs(self):
+        if job_type is None:
+            job_type = ''
+        else:
+            if job_type.lower().strip() not in self.JOB_TYPES:
```
elif off course

Avielyo10 replied:
Exactly, change it to `elif`

Avielyo10 commented on bin/glassdoor israel cities data.json:
```diff
@@ -0,0 +1,1596 @@
+{
```
That's great!`bin` directory should contain a binary files and not data files, just move it to `data` directory and remove `bin`

Avielyo10 reviewed (changes requested):
Great progress!please add some unit test to test your feature!

ofeksr committed:
Delete .gitignore

ofeksr commented on legendary_eureka.py:
```diff
@@ -31,9 +33,120 @@ def get_jobs(self):
-    # TODO
-    def get_jobs(self):
-        pass
+    # TODO: wrap requests.post parts with try / except.
```
@Avielyo10 i would like to know how to do it right, and also how to unittest the except case with connection exception.

Avielyo10 replied:
See [pytest](https://docs.pytest.org/en/latest/)

ofeksr commented on legendary_eureka.py:
```diff
@@ -31,9 +33,120 @@ def get_jobs(self):
 
 class Glassdoor(Resource):
+
+    JOB_TYPES = ['fulltime', 'parttime', 'contract', 'internship', 'temporary', 'apprenticeship', 'entrylevel']
```
@Avielyo10 Should I change to `__JOB_TYPES` ?
As we don't want this list to be changed..

Avielyo10 replied:
That's totally keep it that way

ofeksr committed:
Minor fixes from review and adding unittest for glassdoor.get_jobs function.

ofeksr committed:
Merge remote-tracking branch 'ofeksr/glassdoor' into glassdoor

ofeksr committed:
Delete glassdoor israel cities data.json

Avielyo10 commented on tests/glassdoor_tests.py:
```diff
@@ -0,0 +1,54 @@
+                        self.glassdoor.get_jobs(
+                            job_title='Engineer',
+                            job_location='Miami')
+                    )
```
Replace to switch-case statement

Avielyo10 commented on legendary_eureka.py:
```diff
@@ -31,9 +33,123 @@ def get_jobs(self):
+        # printing jobs
+        for val in job_listings.values():
+            for index, job_info in enumerate(val.values()):
+                print(f'#{index + 1} Job Title: {job_info[0]}\nJob URL: {job_info[1]}')
```
We don't need to print this to stdout, the function can return a dictionary and then we will create a function that prints  dictionaries

ofeksr committed:
Unittest sub tests switched to "switch-case statement".
Printing results changed to return dictionary.

ofeksr committed:
Merge remote-tracking branch 'ofeksr/glassdoor' into glassdoor

Avielyo10 commented on tests/glassdoor_tests.py:
```diff
@@ -0,0 +1,68 @@
+        Test all get_jobs function with all possible parameters.
+        """
+
+        def zero():
```
Meaningful names please :)

Avielyo10 reviewed (changes requested):
Meaningful names for all of the tests, great jobs!

ofeksr committed:
Adding meaningful names for sub tests.

Avielyo10 committed:
Merge branch 'master' into glassdoor
